# Lab | Customer Analysis Round 1

### Remember the process:

1. Case Study
2. Get data
3. Cleaning/Wrangling/EDA
4. Processing Data
5. Modeling
6. Validation
7. Reporting

### Abstract

The objective of this data is to understand customer demographics and buying behavior. 

For this lab, we will gather the data from 3 _csv_ files that are provided in the `Data` folder. Use that data and complete the data cleaning tasks as mentioned later in the instructions.


### Instructions

- Read the three files into python as dataframes
- Show the DataFrame's shape.
- Standardize header names.
- Rearrange the columns in the dataframe as needed
- Concatenate the three dataframes
- Which columns are numerical?
- Which columns are categorical?
- Understand the meaning of all columns
- Perform the data cleaning operations mentioned so far in class

  - Delete the column education and the number of open complaints from the dataframe.
  - Correct the values in the column customer lifetime value. They are given as a percent, so multiply them by 100 and change `dtype` to `numerical` type.
  - Check for duplicate rows in the data and remove if any.
  - Filter out the data for customers who have an income of 0 or less.
  - Check and deal with `NaN` values.
  
- Datetime format - Extract the months from the dataset and store in a separate column. Then filter the data to show only the information for the first quarter , ie. January, February and March. _Hint_: If data from March does not exist, consider only January and February.
- BONUS: Put all the previously mentioned data transformations into a function."""

## My code

### 1. Read the three files into python as dataframes

In [3]:
import pandas as pd  
import numpy as np

In [9]:
file1 = pd.read_csv("Data/file1.csv")
file2 = pd.read_csv("Data/file2.csv")
file3 = pd.read_csv("Data/file3.csv")

In [10]:
#Vamos a ver las 5 primeras lineas del file1
file1.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [11]:
#Vamos a ver las 5 primeras lineas del file2
file2.head()

Customer          ST GENDER Education Customer Lifetime Value  Income  \
0  GS98873     Arizona      F  Bachelor              323912.47%   16061   
1  CW49887  California      F    Master              462680.11%   79487   
2  MY31220  California      F   College              899704.02%   54230   
3  UH35128      Oregon      F   College             2580706.30%   71210   
4  WH52799     Arizona      F   College              380812.21%   94903   

   Monthly Premium Auto Number of Open Complaints  Total Claim Amount  \
0                    88                    1/0/00               633.6   
1                   114                    1/0/00               547.2   
2                   112                    1/0/00               537.6   
3                   214                    1/1/00              1027.2   
4                    94                    1/0/00               451.2   

      Policy Type  Vehicle Class  
0   Personal Auto  Four-Door Car  
1    Special Auto            SUV  
2   Personal Auto   Two-Door Car  
3   Personal Auto     Luxury Car  
4  Corporate Auto   Two-Door Car

In [12]:
#Vamos a ver las 5 primeras lineas del file3
file3.head()

Customer       State  Customer Lifetime Value             Education Gender  \
0  SA25987  Washington              3479.137523  High School or Below      M   
1  TB86706     Arizona              2502.637401                Master      M   
2  ZL73902      Nevada              3265.156348              Bachelor      F   
3  KX23516  California              4455.843406  High School or Below      F   
4  FN77294  California              7704.958480  High School or Below      M   

   Income  Monthly Premium Auto  Number of Open Complaints    Policy Type  \
0       0                   104                          0  Personal Auto   
1       0                    66                          0  Personal Auto   
2   25820                    82                          0  Personal Auto   
3       0                   121                          0  Personal Auto   
4   30366                   101                          2  Personal Auto   

   Total Claim Amount  Vehicle Class  
0          499.200000   Two-Door Car  
1            3.468912   Two-Door Car  
2          393.600000  Four-Door Car  
3          699.615192            SUV  
4          484.800000            SUV

### 2. Show the DataFrame's shape.

In [13]:
### file1 shape
print(file1.shape)

(4008, 11)


##### Vemos file1 que tiene 4008 filas y 11 columnas.

In [14]:
### file2 shape
print(file2.shape)

(996, 11)


##### Vemos file2 que tiene 996 filas y 11 columnas.

In [15]:
### file3 shape
print(file3.shape)

(7070, 11)


##### Vemos file3 que tiene 7070 filas y 11 columnas.

### 3. Standardize header names.

##### Para estandarizar los nombres de las columnas vamos a realizar las buenas prácticas de poner todos los encabezados en minúscula y separar el texto con _, lo haremos una función llamada stand_col.

In [16]:
#file1
stand_col = []
for i in range(len(file1.columns)):
    stand_col.append(file1.columns[i].lower().replace(' ','_'))
file1.columns = stand_col

In [17]:
#comprovemos
file1.head()

customer          st gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [18]:
#file2
stand_col = []
for i in range(len(file2.columns)):
    stand_col.append(file2.columns[i].lower().replace(' ','_'))
file2.columns = stand_col

In [19]:
#comprovemos
file2.head()

customer          st gender education customer_lifetime_value  income  \
0  GS98873     Arizona      F  Bachelor              323912.47%   16061   
1  CW49887  California      F    Master              462680.11%   79487   
2  MY31220  California      F   College              899704.02%   54230   
3  UH35128      Oregon      F   College             2580706.30%   71210   
4  WH52799     Arizona      F   College              380812.21%   94903   

   monthly_premium_auto number_of_open_complaints  total_claim_amount  \
0                    88                    1/0/00               633.6   
1                   114                    1/0/00               547.2   
2                   112                    1/0/00               537.6   
3                   214                    1/1/00              1027.2   
4                    94                    1/0/00               451.2   

      policy_type  vehicle_class  
0   Personal Auto  Four-Door Car  
1    Special Auto            SUV  
2   Personal Auto   Two-Door Car  
3   Personal Auto     Luxury Car  
4  Corporate Auto   Two-Door Car

In [20]:
#file3
stand_col = []
for i in range(len(file3.columns)):
    stand_col.append(file3.columns[i].lower().replace(' ','_'))
file3.columns = stand_col

In [21]:
#Comprovemos
file3.head()

customer       state  customer_lifetime_value             education gender  \
0  SA25987  Washington              3479.137523  High School or Below      M   
1  TB86706     Arizona              2502.637401                Master      M   
2  ZL73902      Nevada              3265.156348              Bachelor      F   
3  KX23516  California              4455.843406  High School or Below      F   
4  FN77294  California              7704.958480  High School or Below      M   

   income  monthly_premium_auto  number_of_open_complaints    policy_type  \
0       0                   104                          0  Personal Auto   
1       0                    66                          0  Personal Auto   
2   25820                    82                          0  Personal Auto   
3       0                   121                          0  Personal Auto   
4   30366                   101                          2  Personal Auto   

   total_claim_amount  vehicle_class  
0          499.200000   Two-Door Car  
1            3.468912   Two-Door Car  
2          393.600000  Four-Door Car  
3          699.615192            SUV  
4          484.800000            SUV

##### Vamos a unificar los nombres de las columnas, ya que vemos que en file1 y en file2 tenemos st y en file3 state, renombramos en file1 y file2 la columna st a state.

In [22]:
#file1
file1 = file1.rename(columns={'st':'state'})

In [23]:
#Comprovamos
file1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   1071 non-null   object 
 1   state                      1071 non-null   object 
 2   gender                     954 non-null    object 
 3   education                  1071 non-null   object 
 4   customer_lifetime_value    1068 non-null   object 
 5   income                     1071 non-null   float64
 6   monthly_premium_auto       1071 non-null   float64
 7   number_of_open_complaints  1071 non-null   object 
 8   policy_type                1071 non-null   object 
 9   vehicle_class              1071 non-null   object 
 10  total_claim_amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [24]:
#file2
file2 = file2.rename(columns={'st':'state'})

In [25]:
file2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 996 entries, 0 to 995
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   996 non-null    object 
 1   state                      996 non-null    object 
 2   gender                     991 non-null    object 
 3   education                  996 non-null    object 
 4   customer_lifetime_value    992 non-null    object 
 5   income                     996 non-null    int64  
 6   monthly_premium_auto       996 non-null    int64  
 7   number_of_open_complaints  996 non-null    object 
 8   total_claim_amount         996 non-null    float64
 9   policy_type                996 non-null    object 
 10  vehicle_class              996 non-null    object 
dtypes: float64(1), int64(2), object(8)
memory usage: 85.7+ KB


### 4. Rearrange the columns in the dataframe as needed.

In [26]:
print(file1.info())
print(file2.info())
print(file3.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   1071 non-null   object 
 1   state                      1071 non-null   object 
 2   gender                     954 non-null    object 
 3   education                  1071 non-null   object 
 4   customer_lifetime_value    1068 non-null   object 
 5   income                     1071 non-null   float64
 6   monthly_premium_auto       1071 non-null   float64
 7   number_of_open_complaints  1071 non-null   object 
 8   policy_type                1071 non-null   object 
 9   vehicle_class              1071 non-null   object 
 10  total_claim_amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 996 entries, 0 to 995
Data columns (total 11 columns):
 

##### Vamos a realizar un print de los 3 archivos y vemos que las columnas no mantienen el mismo orden, vamos a ello. Usaremos file1 como referencia para tener las columnas en el mismo orden de file2 y file3.

In [27]:
file2 = file2[[
'customer', 'state', 'gender', 'education', 'customer_lifetime_value', 'income', 'monthly_premium_auto', 'number_of_open_complaints', 'policy_type', 'vehicle_class', 'total_claim_amount']]

In [28]:
#Comprovemos
file2.head(1)

customer    state gender education customer_lifetime_value  income  \
0  GS98873  Arizona      F  Bachelor              323912.47%   16061   

   monthly_premium_auto number_of_open_complaints    policy_type  \
0                    88                    1/0/00  Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car               633.6

In [29]:
file3 = file3[[
'customer', 'state', 'gender', 'education', 'customer_lifetime_value', 'income', 'monthly_premium_auto', 'number_of_open_complaints', 'policy_type', 'vehicle_class', 'total_claim_amount']]

In [30]:
#Comprovemos
file3.head(1)

customer       state gender             education  customer_lifetime_value  \
0  SA25987  Washington      M  High School or Below              3479.137523   

   income  monthly_premium_auto  number_of_open_complaints    policy_type  \
0       0                   104                          0  Personal Auto   

  vehicle_class  total_claim_amount  
0  Two-Door Car               499.2

### 5. Concatenate the three dataframes.

In [31]:
data = pd.concat([file1, file2, file3], axis=0, ignore_index=True)

In [32]:
data

customer       state gender             education  \
0      RB50392  Washington    NaN                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
12069  LA72316  California      M              Bachelor   
12070  PK87824  California      F               College   
12071  TD14365  California      M              Bachelor   
12072  UP19263  California      M               College   
12073  Y167826  California      M               College   

      customer_lifetime_value   income  monthly_premium_auto  \
0                         NaN      0.0                1000.0   
1                  697953.59%      0.0                  94.0   
2                 1288743.17%  48767.0                 108.0   
3                  764586.18%      0.0                 106.0   
4                  536307.65%  36357.0                  68.0   
...                       ...      ...                   ...   
12069             23405.98798  71941.0                  73.0   
12070             3096.511217  21604.0                  79.0   
12071             8163.890428      0.0                  85.0   
12072             7524.442436  21941.0                  96.0   
12073             2611.836866      0.0                  77.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                        1/0/00   Personal Auto  Four-Door Car   
1                        1/0/00   Personal Auto  Four-Door Car   
2                        1/0/00   Personal Auto   Two-Door Car   
3                        1/0/00  Corporate Auto            SUV   
4                        1/0/00   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
12069                         0   Personal Auto  Four-Door Car   
12070                         0  Corporate Auto  Four-Door Car   
12071                         3  Corporate Auto  Four-Door Car   
12072                         0   Personal Auto  Four-Door Car   
12073                         0  Corporate Auto   Two-Door Car   

       total_claim_amount  
0                2.704934  
1             1131.464935  
2              566.472247  
3              529.881344  
4               17.269323  
...                   ...  
12069          198.234764  
12070          379.200000  
12071          790.784983  
12072          691.200000  
12073          369.600000  

[12074 rows x 11 columns]

In [33]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12074 entries, 0 to 12073
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customer                   9137 non-null   object 
 1   state                      9137 non-null   object 
 2   gender                     9015 non-null   object 
 3   education                  9137 non-null   object 
 4   customer_lifetime_value    9130 non-null   object 
 5   income                     9137 non-null   float64
 6   monthly_premium_auto       9137 non-null   float64
 7   number_of_open_complaints  9137 non-null   object 
 8   policy_type                9137 non-null   object 
 9   vehicle_class              9137 non-null   object 
 10  total_claim_amount         9137 non-null   float64
dtypes: float64(3), object(8)
memory usage: 1.0+ MB


### 6. Which columns are numerical?

In [43]:
#Para saber que columnas son numéricas podemos usar data.types
data.dtypes

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value       object
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

In [54]:
#Con get_numeric_data podemos saber que columnas son numericas
data._get_numeric_data()

income  monthly_premium_auto  total_claim_amount
0          0.0                1000.0            2.704934
1          0.0                  94.0         1131.464935
2      48767.0                 108.0          566.472247
3          0.0                 106.0          529.881344
4      36357.0                  68.0           17.269323
...        ...                   ...                 ...
12069  71941.0                  73.0          198.234764
12070  21604.0                  79.0          379.200000
12071      0.0                  85.0          790.784983
12072  21941.0                  96.0          691.200000
12073      0.0                  77.0          369.600000

[12074 rows x 3 columns]

### 7. Which columns are categorical?

In [42]:
data.dtypes

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value       object
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

In [57]:
#Columnas categóricas
columnas_categoricas = data.select_dtypes(include=['object', 'category'])
columnas_categoricas

customer       state gender             education  \
0      RB50392  Washington    NaN                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
12069  LA72316  California      M              Bachelor   
12070  PK87824  California      F               College   
12071  TD14365  California      M              Bachelor   
12072  UP19263  California      M               College   
12073  Y167826  California      M               College   

      customer_lifetime_value number_of_open_complaints     policy_type  \
0                         NaN                    1/0/00   Personal Auto   
1                  697953.59%                    1/0/00   Personal Auto   
2                 1288743.17%                    1/0/00   Personal Auto   
3                  764586.18%                    1/0/00  Corporate Auto   
4                  536307.65%                    1/0/00   Personal Auto   
...                       ...                       ...             ...   
12069             23405.98798                         0   Personal Auto   
12070             3096.511217                         0  Corporate Auto   
12071             8163.890428                         3  Corporate Auto   
12072             7524.442436                         0   Personal Auto   
12073             2611.836866                         0  Corporate Auto   

       vehicle_class  
0      Four-Door Car  
1      Four-Door Car  
2       Two-Door Car  
3                SUV  
4      Four-Door Car  
...              ...  
12069  Four-Door Car  
12070  Four-Door Car  
12071  Four-Door Car  
12072  Four-Door Car  
12073   Two-Door Car  

[12074 rows x 8 columns]

### 8. Understand the meaning of all columns

##### Todo y que ya hemos usado algunos de los comonados vamos a realizar una inspección paso a paso.

In [60]:
#Vemos los 5 primeros registros del data y sus columnas
data.head()

customer       state gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [61]:
#Vemos los 5 priúltimos registros del data y sus columnas
data.tail()

customer       state gender education customer_lifetime_value   income  \
12069  LA72316  California      M  Bachelor             23405.98798  71941.0   
12070  PK87824  California      F   College             3096.511217  21604.0   
12071  TD14365  California      M  Bachelor             8163.890428      0.0   
12072  UP19263  California      M   College             7524.442436  21941.0   
12073  Y167826  California      M   College             2611.836866      0.0   

       monthly_premium_auto number_of_open_complaints     policy_type  \
12069                  73.0                         0   Personal Auto   
12070                  79.0                         0  Corporate Auto   
12071                  85.0                         3  Corporate Auto   
12072                  96.0                         0   Personal Auto   
12073                  77.0                         0  Corporate Auto   

       vehicle_class  total_claim_amount  
12069  Four-Door Car          198.234764  
12070  Four-Door Car          379.200000  
12071  Four-Door Car          790.784983  
12072  Four-Door Car          691.200000  
12073   Two-Door Car          369.600000

In [64]:
#Vamos a ver un registro aleatorio.
data.sample()

customer   state gender education customer_lifetime_value   income  \
10701  GC40999  Oregon      M   College             11790.49617  25251.0   

       monthly_premium_auto number_of_open_complaints    policy_type  \
10701                  66.0                         0  Personal Auto   

       vehicle_class  total_claim_amount  
10701  Four-Door Car               316.8

In [66]:
#Vamos a ver un resumen estadístoco de las columnas categorias 
data.describe()

income  monthly_premium_auto  total_claim_amount
count   9137.000000           9137.000000         9137.000000
mean   37828.820291            110.391266          430.527140
std    30358.716159            581.376032          289.582968
min        0.000000             61.000000            0.099007
25%        0.000000             68.000000          266.996814
50%    34244.000000             83.000000          377.561463
75%    62447.000000            109.000000          546.420009
max    99981.000000          35354.000000         2893.239678

In [74]:
data.dtypes

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value       object
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

#### Que significa cada columna:
- customer, es el id de cada registro
- state, estado donde vive
- gender, genero de la persona
- education, educación recibida
- customer_lifetime_value, ganancias que puede generar un cliente
- income, ingreso
- monthly_premium_auto, prima mensual del seguro
- number_of_open_complaints, número de quejas abiertas o en proceso
- policy_type, es el tipo de póliza
- vehicle_class, tipo/modelo de vehículo
- total_claim_amount, cantidad total reclamada de una póliza.

### 9. Perform the data cleaning operations mentioned so far in class.

In [75]:
#Vamos a realizar la limpieza del data
# Primero corregimos los tipos de datos, vamos a pasar a numérico la columna 'customer_lifetime_value':
data['customer_lifetime_value'] = pd.to_numeric(data['customer_lifetime_value'], errors='coerce')## lo pasamos a numerico y los errores coerce

In [76]:
#Lo mismo para la columna number_of_open_complaints
data['number_of_open_complaints'] =  pd.to_numeric(data['number_of_open_complaints'], errors='coerce')

In [78]:
#Comprovamos que 'customer_lifetime_value' y 'number_of_open_complaints' estan ahora en la lista de datos numéricos.
data._get_numeric_data()

customer_lifetime_value   income  monthly_premium_auto  \
0                          NaN      0.0                1000.0   
1                          NaN      0.0                  94.0   
2                          NaN  48767.0                 108.0   
3                          NaN      0.0                 106.0   
4                          NaN  36357.0                  68.0   
...                        ...      ...                   ...   
12069             23405.987980  71941.0                  73.0   
12070              3096.511217  21604.0                  79.0   
12071              8163.890428      0.0                  85.0   
12072              7524.442436  21941.0                  96.0   
12073              2611.836866      0.0                  77.0   

       number_of_open_complaints  total_claim_amount  
0                            NaN            2.704934  
1                            NaN         1131.464935  
2                            NaN          566.472247  
3                            NaN          529.881344  
4                            NaN           17.269323  
...                          ...                 ...  
12069                        0.0          198.234764  
12070                        0.0          379.200000  
12071                        3.0          790.784983  
12072                        0.0          691.200000  
12073                        0.0          369.600000  

[12074 rows x 5 columns]

In [79]:
#vamos a eleminar los duplicados.
data_clean = data.drop_duplicates()

In [80]:
data_clean

customer       state gender             education  \
0      RB50392  Washington    NaN                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
12069  LA72316  California      M              Bachelor   
12070  PK87824  California      F               College   
12071  TD14365  California      M              Bachelor   
12072  UP19263  California      M               College   
12073  Y167826  California      M               College   

       customer_lifetime_value   income  monthly_premium_auto  \
0                          NaN      0.0                1000.0   
1                          NaN      0.0                  94.0   
2                          NaN  48767.0                 108.0   
3                          NaN      0.0                 106.0   
4                          NaN  36357.0                  68.0   
...                        ...      ...                   ...   
12069             23405.987980  71941.0                  73.0   
12070              3096.511217  21604.0                  79.0   
12071              8163.890428      0.0                  85.0   
12072              7524.442436  21941.0                  96.0   
12073              2611.836866      0.0                  77.0   

       number_of_open_complaints     policy_type  vehicle_class  \
0                            NaN   Personal Auto  Four-Door Car   
1                            NaN   Personal Auto  Four-Door Car   
2                            NaN   Personal Auto   Two-Door Car   
3                            NaN  Corporate Auto            SUV   
4                            NaN   Personal Auto  Four-Door Car   
...                          ...             ...            ...   
12069                        0.0   Personal Auto  Four-Door Car   
12070                        0.0  Corporate Auto  Four-Door Car   
12071                        3.0  Corporate Auto  Four-Door Car   
12072                        0.0   Personal Auto  Four-Door Car   
12073                        0.0  Corporate Auto   Two-Door Car   

       total_claim_amount  
0                2.704934  
1             1131.464935  
2              566.472247  
3              529.881344  
4               17.269323  
...                   ...  
12069          198.234764  
12070          379.200000  
12071          790.784983  
12072          691.200000  
12073          369.600000  

[9135 rows x 11 columns]

In [82]:
#Vamos a ver el procentaje de nulos de cada columna:
nulls_df = pd.DataFrame(round(data_clean.isna().sum()/len(data_clean),4)*100)
nulls_df

0
customer                    0.01
state                       0.01
gender                      1.35
education                   0.01
customer_lifetime_value    22.61
income                      0.01
monthly_premium_auto        0.01
number_of_open_complaints  22.61
policy_type                 0.01
vehicle_class               0.01
total_claim_amount          0.01

In [83]:
nulls_df = nulls_df.reset_index()
nulls_df

index      0
0                    customer   0.01
1                       state   0.01
2                      gender   1.35
3                   education   0.01
4     customer_lifetime_value  22.61
5                      income   0.01
6        monthly_premium_auto   0.01
7   number_of_open_complaints  22.61
8                 policy_type   0.01
9               vehicle_class   0.01
10         total_claim_amount   0.01

In [84]:
nulls_df.columns = ['header_name', 'percent_nulls']
nulls_df

header_name  percent_nulls
0                    customer           0.01
1                       state           0.01
2                      gender           1.35
3                   education           0.01
4     customer_lifetime_value          22.61
5                      income           0.01
6        monthly_premium_auto           0.01
7   number_of_open_complaints          22.61
8                 policy_type           0.01
9               vehicle_class           0.01
10         total_claim_amount           0.01

In [85]:
#Vamos a ver las columnas que tienen un porcentage de nulos mayor al 3%. Todo y que a simple vista se puede observar.
columns_drop = nulls_df[nulls_df['percent_nulls']>3]['header_name']
print(columns_drop.values)

['customer_lifetime_value' 'number_of_open_complaints']


In [86]:
## Vamos a imputar los valores nulos en customer_lifetime_value con el valor promedio
mean_customer_lifetime_value = np.mean(data_clean['customer_lifetime_value'])
data_clean['customer_lifetime_value'] = data_clean['customer_lifetime_value'].fillna(mean_customer_lifetime_value)

/var/folders/zr/v_qv79zn4xv2dhqwl7djvy340000gn/T/ipykernel_4600/2145751722.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_clean['customer_lifetime_value'] = data_clean['customer_lifetime_value'].fillna(mean_customer_lifetime_value)


In [87]:
data_clean

customer       state gender             education  \
0      RB50392  Washington    NaN                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
12069  LA72316  California      M              Bachelor   
12070  PK87824  California      F               College   
12071  TD14365  California      M              Bachelor   
12072  UP19263  California      M               College   
12073  Y167826  California      M               College   

       customer_lifetime_value   income  monthly_premium_auto  \
0                  8028.803803      0.0                1000.0   
1                  8028.803803      0.0                  94.0   
2                  8028.803803  48767.0                 108.0   
3                  8028.803803      0.0                 106.0   
4                  8028.803803  36357.0                  68.0   
...                        ...      ...                   ...   
12069             23405.987980  71941.0                  73.0   
12070              3096.511217  21604.0                  79.0   
12071              8163.890428      0.0                  85.0   
12072              7524.442436  21941.0                  96.0   
12073              2611.836866      0.0                  77.0   

       number_of_open_complaints     policy_type  vehicle_class  \
0                            NaN   Personal Auto  Four-Door Car   
1                            NaN   Personal Auto  Four-Door Car   
2                            NaN   Personal Auto   Two-Door Car   
3                            NaN  Corporate Auto            SUV   
4                            NaN   Personal Auto  Four-Door Car   
...                          ...             ...            ...   
12069                        0.0   Personal Auto  Four-Door Car   
12070                        0.0  Corporate Auto  Four-Door Car   
12071                        3.0  Corporate Auto  Four-Door Car   
12072                        0.0   Personal Auto  Four-Door Car   
12073                        0.0  Corporate Auto   Two-Door Car   

       total_claim_amount  
0                2.704934  
1             1131.464935  
2              566.472247  
3              529.881344  
4               17.269323  
...                   ...  
12069          198.234764  
12070          379.200000  
12071          790.784983  
12072          691.200000  
12073          369.600000  

[9135 rows x 11 columns]

In [88]:
## Vamos a imputar los valores nulos en number_of_open_complaints con el valor 0.0 porque lo vamos a consider como que no hay quejas abiertas.
data_clean['number_of_open_complaints'] = data_clean['number_of_open_complaints'].fillna(0.0)

/var/folders/zr/v_qv79zn4xv2dhqwl7djvy340000gn/T/ipykernel_4600/960584678.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_clean['number_of_open_complaints'] = data_clean['number_of_open_complaints'].fillna(0.0)


In [89]:
data_clean

customer       state gender             education  \
0      RB50392  Washington    NaN                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
12069  LA72316  California      M              Bachelor   
12070  PK87824  California      F               College   
12071  TD14365  California      M              Bachelor   
12072  UP19263  California      M               College   
12073  Y167826  California      M               College   

       customer_lifetime_value   income  monthly_premium_auto  \
0                  8028.803803      0.0                1000.0   
1                  8028.803803      0.0                  94.0   
2                  8028.803803  48767.0                 108.0   
3                  8028.803803      0.0                 106.0   
4                  8028.803803  36357.0                  68.0   
...                        ...      ...                   ...   
12069             23405.987980  71941.0                  73.0   
12070              3096.511217  21604.0                  79.0   
12071              8163.890428      0.0                  85.0   
12072              7524.442436  21941.0                  96.0   
12073              2611.836866      0.0                  77.0   

       number_of_open_complaints     policy_type  vehicle_class  \
0                            0.0   Personal Auto  Four-Door Car   
1                            0.0   Personal Auto  Four-Door Car   
2                            0.0   Personal Auto   Two-Door Car   
3                            0.0  Corporate Auto            SUV   
4                            0.0   Personal Auto  Four-Door Car   
...                          ...             ...            ...   
12069                        0.0   Personal Auto  Four-Door Car   
12070                        0.0  Corporate Auto  Four-Door Car   
12071                        3.0  Corporate Auto  Four-Door Car   
12072                        0.0   Personal Auto  Four-Door Car   
12073                        0.0  Corporate Auto   Two-Door Car   

       total_claim_amount  
0                2.704934  
1             1131.464935  
2              566.472247  
3              529.881344  
4               17.269323  
...                   ...  
12069          198.234764  
12070          379.200000  
12071          790.784983  
12072          691.200000  
12073          369.600000  

[9135 rows x 11 columns]

In [92]:
#Vamos a ver el procentaje de nulos de cada columna:
nulls_df = pd.DataFrame(round(data_clean.isna().sum()/len(data_clean),4)*100)
nulls_df

0
customer                   0.01
state                      0.01
gender                     1.35
education                  0.01
customer_lifetime_value    0.00
income                     0.01
monthly_premium_auto       0.01
number_of_open_complaints  0.00
policy_type                0.01
vehicle_class              0.01
total_claim_amount         0.01

  - Delete the column education and the number of open complaints from the dataframe.

In [118]:
data_clean = data_clean.drop (['education', 'number_of_open_complaints'], axis=1)

KeyError: "['education', 'number_of_open_complaints'] not found in axis"

In [97]:
data_clean

customer       state gender  customer_lifetime_value   income  \
0      RB50392  Washington    NaN              8028.803803      0.0   
1      QZ44356     Arizona      F              8028.803803      0.0   
2      AI49188      Nevada      F              8028.803803  48767.0   
3      WW63253  California      M              8028.803803      0.0   
4      GA49547  Washington      M              8028.803803  36357.0   
...        ...         ...    ...                      ...      ...   
12069  LA72316  California      M             23405.987980  71941.0   
12070  PK87824  California      F              3096.511217  21604.0   
12071  TD14365  California      M              8163.890428      0.0   
12072  UP19263  California      M              7524.442436  21941.0   
12073  Y167826  California      M              2611.836866      0.0   

       monthly_premium_auto     policy_type  vehicle_class  total_claim_amount  
0                    1000.0   Personal Auto  Four-Door Car            2.704934  
1                      94.0   Personal Auto  Four-Door Car         1131.464935  
2                     108.0   Personal Auto   Two-Door Car          566.472247  
3                     106.0  Corporate Auto            SUV          529.881344  
4                      68.0   Personal Auto  Four-Door Car           17.269323  
...                     ...             ...            ...                 ...  
12069                  73.0   Personal Auto  Four-Door Car          198.234764  
12070                  79.0  Corporate Auto  Four-Door Car          379.200000  
12071                  85.0  Corporate Auto  Four-Door Car          790.784983  
12072                  96.0   Personal Auto  Four-Door Car          691.200000  
12073                  77.0  Corporate Auto   Two-Door Car          369.600000  

[9135 rows x 9 columns]

  - Correct the values in the column customer lifetime value. They are given as a percent, so multiply them by 100 and change `dtype` to `numerical` type.

In [98]:
data_clean ['customer_lifetime_value'] = data_clean ['customer_lifetime_value']*100

In [101]:
data_clean ['customer_lifetime_value'] = pd.to_numeric(data_clean ['customer_lifetime_value'])

In [102]:
data_clean

customer       state gender  customer_lifetime_value   income  \
0      RB50392  Washington    NaN             8.028804e+05      0.0   
1      QZ44356     Arizona      F             8.028804e+05      0.0   
2      AI49188      Nevada      F             8.028804e+05  48767.0   
3      WW63253  California      M             8.028804e+05      0.0   
4      GA49547  Washington      M             8.028804e+05  36357.0   
...        ...         ...    ...                      ...      ...   
12069  LA72316  California      M             2.340599e+06  71941.0   
12070  PK87824  California      F             3.096511e+05  21604.0   
12071  TD14365  California      M             8.163890e+05      0.0   
12072  UP19263  California      M             7.524442e+05  21941.0   
12073  Y167826  California      M             2.611837e+05      0.0   

       monthly_premium_auto     policy_type  vehicle_class  total_claim_amount  
0                    1000.0   Personal Auto  Four-Door Car            2.704934  
1                      94.0   Personal Auto  Four-Door Car         1131.464935  
2                     108.0   Personal Auto   Two-Door Car          566.472247  
3                     106.0  Corporate Auto            SUV          529.881344  
4                      68.0   Personal Auto  Four-Door Car           17.269323  
...                     ...             ...            ...                 ...  
12069                  73.0   Personal Auto  Four-Door Car          198.234764  
12070                  79.0  Corporate Auto  Four-Door Car          379.200000  
12071                  85.0  Corporate Auto  Four-Door Car          790.784983  
12072                  96.0   Personal Auto  Four-Door Car          691.200000  
12073                  77.0  Corporate Auto   Two-Door Car          369.600000  

[9135 rows x 9 columns]

  - Check for duplicate rows in the data and remove if any.

In [104]:
#Este proceso lo he hecho anteriormente.
data_clean = data_clean.drop_duplicates()

  - Filter out the data for customers who have an income of 0 or less.

In [106]:
data_clean_filtered = data_clean[data_clean['income'] > 0]
data_clean_filtered

customer       state gender  customer_lifetime_value   income  \
2      AI49188      Nevada      F             8.028804e+05  48767.0   
4      GA49547  Washington      M             8.028804e+05  36357.0   
5      OC83172      Oregon      F             8.028804e+05  62902.0   
6      XZ87318      Oregon      F             8.028804e+05  55350.0   
8      DY87989      Oregon      M             8.028804e+05  14072.0   
...        ...         ...    ...                      ...      ...   
12067  TF56202  California      M             5.032165e+05  66367.0   
12068  YM19146  California      F             4.100399e+05  47761.0   
12069  LA72316  California      M             2.340599e+06  71941.0   
12070  PK87824  California      F             3.096511e+05  21604.0   
12072  UP19263  California      M             7.524442e+05  21941.0   

       monthly_premium_auto     policy_type  vehicle_class  total_claim_amount  
2                     108.0   Personal Auto   Two-Door Car          566.472247  
4                      68.0   Personal Auto  Four-Door Car           17.269323  
5                      69.0   Personal Auto   Two-Door Car          159.383042  
6                      67.0  Corporate Auto  Four-Door Car          321.600000  
8                      71.0  Corporate Auto  Four-Door Car          511.200000  
...                     ...             ...            ...                 ...  
12067                  64.0   Personal Auto   Two-Door Car          307.200000  
12068                 104.0   Personal Auto  Four-Door Car          541.282007  
12069                  73.0   Personal Auto  Four-Door Car          198.234764  
12070                  79.0  Corporate Auto  Four-Door Car          379.200000  
12072                  96.0   Personal Auto  Four-Door Car          691.200000  

[6840 rows x 9 columns]

  - Check and deal with `NaN` values.

In [107]:
#Este proceso lo he hecho anteiormente.
nulls_df = pd.DataFrame(round(data_clean.isna().sum()/len(data_clean),4)*100)
nulls_df

0
customer                 0.01
state                    0.01
gender                   1.35
customer_lifetime_value  0.00
income                   0.01
monthly_premium_auto     0.01
policy_type              0.01
vehicle_class            0.01
total_claim_amount       0.01

He dejado los porcentajes de nulos por debajo del 3%, no se si era lo que se pedia o tenemos que dejar el data limpio al 100%

### 10. Datetime format - Extract the months from the dataset and store in a separate column. Then filter the data to show only the information for the first quarter , ie. January, February and March. _Hint_: If data from March does not exist, consider only January and February.

In [108]:
#Para esta pregunta voy a usar el csv marketing_customer_analysis, ya que en el file1, file2, y file no tengo la columna effective_to_date
file = pd.read_csv('Data/marketing_customer_analysis.csv')
file.dtypes

Unnamed: 0                         int64
Customer                          object
State                             object
Customer Lifetime Value          float64
Response                          object
Coverage                          object
Education                         object
Effective To Date                 object
EmploymentStatus                  object
Gender                            object
Income                             int64
Location Code                     object
Marital Status                    object
Monthly Premium Auto               int64
Months Since Last Claim          float64
Months Since Policy Inception      int64
Number of Open Complaints        float64
Number of Policies                 int64
Policy Type                       object
Policy                            object
Renew Offer Type                  object
Sales Channel                     object
Total Claim Amount               float64
Vehicle Class                     object
Vehicle Size    

In [110]:
#Primero vamos a convertir la columna effective_to_date a formato datetime
file['Effective To Date'] = pd.to_datetime(file['Effective To Date'])

/var/folders/zr/v_qv79zn4xv2dhqwl7djvy340000gn/T/ipykernel_4600/1189664722.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  file['Effective To Date'] = pd.to_datetime(file['Effective To Date'])


In [114]:
#Ahora vamos a extraer el mes y guardarlo en una columna
file['month'] = file['Effective To Date'].dt.month
file.head()

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College        2011-02-18         Employed      M  ...   
1     Basic   College        2011-01-18       Unemployed      F  ...   
2     Basic  Bachelor        2011-02-10         Employed      M  ...   
3  Extended   College        2011-01-11         Employed      M  ...   
4   Premium  Bachelor        2011-01-17    Medical Leave      F  ...   

   Number of Policies     Policy Type        Policy  Renew Offer Type  \
0                   9  Corporate Auto  Corporate L3            Offer3   
1                   1   Personal Auto   Personal L3            Offer4   
2                   2   Personal Auto   Personal L3            Offer3   
3                   2  Corporate Auto  Corporate L3            Offer2   
4                   7   Personal Auto   Personal L2            Offer1   

   Sales Channel  Total Claim Amount  Vehicle Class  Vehicle Size  \
0          Agent          292.800000  Four-Door Car       Medsize   
1    Call Center          744.924331  Four-Door Car       Medsize   
2    Call Center          480.000000            SUV       Medsize   
3         Branch          484.013411  Four-Door Car       Medsize   
4         Branch          707.925645  Four-Door Car       Medsize   

  Vehicle Type month  
0          NaN     2  
1          NaN     1  
2            A     2  
3            A     1  
4          NaN     1  

[5 rows x 27 columns]

In [116]:
#Vamos a filtar por los meses de enero, febrero y marzo

file_primer_trim = file[file['month'].isin([1,2,3])]
file_primer_trim

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0         Basic   College        2011-02-18         Employed      M  ...   
1         Basic   College        2011-01-18       Unemployed      F  ...   
2         Basic  Bachelor        2011-02-10         Employed      M  ...   
3      Extended   College        2011-01-11         Employed      M  ...   
4       Premium  Bachelor        2011-01-17    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor        2011-01-19       Unemployed      F  ...   
10906     Basic   College        2011-01-06         Employed      F  ...   
10907  Extended  Bachelor        2011-02-06         Employed      F  ...   
10908   Premium   College        2011-02-13         Employed      F  ...   
10909     Basic  Bachelor        2011-01-08       Unemployed      M  ...   

       Number of Policies     Policy Type        Policy  Renew Offer Type  \
0                       9  Corporate Auto  Corporate L3            Offer3   
1                       1   Personal Auto   Personal L3            Offer4   
2                       2   Personal Auto   Personal L3            Offer3   
3                       2  Corporate Auto  Corporate L3            Offer2   
4                       7   Personal Auto   Personal L2            Offer1   
...                   ...             ...           ...               ...   
10905                   7   Personal Auto   Personal L1            Offer3   
10906                   6   Personal Auto   Personal L3            Offer2   
10907                   2  Corporate Auto  Corporate L3            Offer1   
10908                   6   Personal Auto   Personal L1            Offer1   
10909                   3   Personal Auto   Personal L1            Offer4   

       Sales Channel  Total Claim Amount  Vehicle Class  Vehicle Size  \
0              Agent          292.800000  Four-Door Car       Medsize   
1        Call Center          744.924331  Four-Door Car       Medsize   
2        Call Center          480.000000            SUV       Medsize   
3             Branch          484.013411  Four-Door Car       Medsize   
4             Branch          707.925645  Four-Door Car       Medsize   
...              ...                 ...            ...           ...   
10905            Web         1214.400000     Luxury Car       Medsize   
10906         Branch          273.018929  Four-Door Car       Medsize   
10907            Web          381.306996     Luxury SUV       Medsize   
10908         Branch          618.288849            SUV       Medsize   
10909            Web         1021.719397            SUV       Medsize   

      Vehicle Type month  
0              NaN     2  
1              NaN     1  
2                A     2  
3                A     1  
4              NaN     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907          NaN     2  
10908            A     2  
10909          NaN     1  

[10910 rows x 27 columns]

### BONUS: Put all the previously mentioned data transformations into a function.

In [ ]:
def data_transform ("Data/file1.csv", "Data/file2.csv", "Data/file3.csv"):
    
    #Para cargar el DataFrame
    file1 = pd.read_csv("Data/file1.csv")
    file2 = pd.read_csv("Data/file2.csv")
    file3 = pd.read_csv("Data/file3.csv")
    
    #Para "normalizar/buenas practicas" los nombres de las columnas poniendo en minúscula los textos y remplazando los espacios entre palabras por '_'.
    file1.columns = file1.columns.str.lower().str.replace(' ', '_')
    file2.columns = file2.columns.str.lower().str.replace(' ', '_')
    file3.columns = file3.columns.str.lower().str.replace(' ', '_')
    
    #Para modificar el nombre de la columnas
    file1.rename(columns = {'st': 'state'}, inplace=True)
    file2.rename(columns = {'st': 'state'}, inplace=True)
    
    #Para concatenar los DataFrames
    data = pd.concat([file1, file2, file3], ignore_index=True)
    
    #Para eliminar las columnas 'education' y 'number_of_open_complaints'
    data.drop(['education', 'number_of_open_complaints'], axis=1, inplace=True)
    
    #Corregir los valores enn 'customer_liftime_value'
    data['customer_lifetime_value'] = pd.to_numeric(data['customer_lifetime_value'].str.replace('%', ''))/100
    data['customer_lifetime_value'] = data['customer_lifetime_value'].fillna(data['customer_lifetime_value'].mean())
    
    #Eliminamos las filas duplicadas
    data.drop_duplicates(inplace=True)
    
    #Filtramos los clientes con ingreso mayor a 0
    data = data[data['income'] > 0]
    
    #Convertimos 'effective_to_date' a datetime y extraemos el mes
    data['effective_to_date'] = pd.to_datetime(data['effective_to_date'])
    data['month'] = data['effective_to_date'].dt.month
    
    #Filtramos para poder obtener el primer trimestre
    data_trimestre1 = data[data['month'].isin([1, 2, 3])]
    
    return data_trimestre1


#Con esta línea de código llamamos a la función con las respectivas rutas a los archivos en nuestro caso concreto.
data_final = transformar_datos("data/file1.csv", "data/file2.csv", "data/file3.csv")
    